### PR TITLE
[WebDriverBiDi] Add test for `originalOpener` to `browsingContext.created`

### PR DESCRIPTION
**Pull request by Lightning00Blade**
Note both `<a target="<X>" />` and `window.open("","<X>")` result to the same algorithm to get/create navigable - https://html.spec.whatwg.org/#the-rules-for-choosing-a-navigable

### CODE REVIEW

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (Lightning00Blade) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:43]: `undefined` = target=_blank

**Review by OrKoN — APPROVED** [review]
LGTM (let us wait for the spec issue and more tests can be added in a follow-up)

File: webdriver/tests/bidi/browsing_context/__init__.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/__init__.py:21]: In Python we don't use camel-case.

```suggestion
    original_opener=None
```

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:23]: This looks like a left-over.

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:35]:
```suggestion
@pytest.mark.parametrize("type_hint", ["tab", "window"])
@pytest.mark.parametrize("features", ["", "popup", "noopener", "noreferrer"])
async def test_original_opener_window_open(bidi_session, wait_for_event, wait_for_future_safe, subscribe_events, type_hint, features):
```

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]: Why does it not always return a valid context? Is that maybe a bug?

**Review by whimboo — CHANGES_REQUESTED** [review]

**Comment by whimboo** [comment]
We should add a test to `browsingContext.getTree` as well, which then can also contain a test for an iframe (where no opener should be present, right?).

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (Lightning00Blade) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]: `window.open` can return `null` and it depends on the provided params.
image

**Comment by Lightning00Blade** [comment]
This only adds test for the `browsingContext.created`, as the title indicates.
The other test will be other in another PR.

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:43]: Maybe we should use that dedicated target so we won't have to know/check  what the default is?

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]: Then we should add another argument to the parameterize section which specifies if the context is valid or not.

Comment (Lightning00Blade) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]: > Then we should add another argument to the parameterize section which specifies if the context is valid or not.

I don't understand what the request here is.

Comment (OrKoN) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]:
```suggestion
        # We use None here as evaluate not always returns value.
```

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:51]: Specify which feature expects a valid context to be returned and which feature just returns `None`. Then we can also do the appropriate checks for context validity.

**Review by OrKoN — APPROVED** [review]

**Comment by whimboo** [comment]
FYI I'm going to wait with the review until the BiDi spec PR has been approved.

**Comment by OrKoN** [comment]
@whimboo we decided not to make any changes on the spec

**Comment by whimboo** [comment]
> @whimboo we decided not to make any changes on the spec

Ah, so it means that https://github.com/w3c/webdriver-bidi/pull/730 can be closed?

**Comment by OrKoN** [comment]
> > @whimboo we decided not to make any changes on the spec
> 
> Ah, so it means that [w3c/webdriver-bidi#730](https://github.com/w3c/webdriver-bidi/pull/730) can be closed?

yes, I will close it.

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/context_created/original_opener.py:12]:
```suggestion
async def test_original_opener_context_create(bidi_session, wait_for_event, wait_for_future_safe, subscribe_events, type_hint):
```

**Review by whimboo — CHANGES_REQUESTED** [review]
I'm confused by the fact that in some cases, None is returned for the originalOpener. According to the original request, as noted in the inline comment, the opener should always be known for top-level browsing contexts in BiDi clients. Consequently, we asked our platform team to add this information, which was previously unavailable for both web content and privileged code. Has something changed here?

File: webdriver/tests/bidi/browsing_context/context_created/original_opener.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: This context is created using WebDriver BiDi and I think the spec does not define what opener should be for those, resulting in the null opener id. @whimboo do you interpret the spec differenty?

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: From Puppeteer perspective, we do not need an opener here since Puppeteer itself opened the context.

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: When using the `browsingContext.create` command with `noopener` to create a new tab/window, it's similar to executing a script like `window.open(..., "noopener")`. In both scenarios, the web page doesn't know about the opener, but privileged code for WebDriver should still be able to return the `opener`. There should be no difference in how the window was opened, and it would be more unusual if we observed different payloads in these cases.

But lets see what @jgraham thinks about.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: @whimboo so which navigable would Firefox use as an opener in this case?

Comment (jgraham) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: This seems like a spec bug. 

We run https://html.spec.whatwg.org/multipage/nav-history-apis.html#window-open-steps which say:
> Let sourceDocument be the [entry global object](https://html.spec.whatwg.org/multipage/webappapis.html#entry-global-object)'s [associated Document](https://html.spec.whatwg.org/multipage/nav-history-apis.html#concept-document-window).

Following "entry global object"'s definition we end up in https://html.spec.whatwg.org/multipage/webappapis.html#entry-execution-context which assumes that there's at least one item in the [JavaScript execution context stack](https://tc39.es/ecma262/#execution-context-stack). However since WebDriver-BiDi is not in fact calling from JS, that isn't true and we have a spec bug. 

I think this was always wrong, but previously we didn't hit any cases where `sourceDocument` was actually used, so it was more or less invisible.

I wonder if we should just update WebDriver to call https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-top-level-traversable directly. In that case we could set the opener to whatever we want, but the obvious options would either be `null` or the (optional) reference context. I think the reference context makes slightly more sense theoretically, but I don't really know what the use cases require.

Comment (Lightning00Blade) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: @whimboo Can you unblock this PR and we can follow up on the spec side.

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/context_created/original_opener.py:24]: We discussed on Matrix and it looks actually fine.

**Review by whimboo — APPROVED** [review]

**Comment by whimboo** [comment]
@lutien if you could take care of the implementation of the `originalOpener` in our implementation that would be great. Due to these changes a lot of wdspec tests will start to fail because the assertions require this field.

**Pull request merged**
